### ISSUE TITLE
binary_accuracy returns wrong value

### ISSUE THREAD

**Issue by kutoga** [fchollet/keras#6540]
Hello

I have two combined model which I train. The output is always a binary number, because I do some kind of classification. For this reason i use binary_crossentropy for the loss and binary_accuracy as a metric.

Somewhen I do some training:

                print("Train model: {}".format(model.name))
                hist = model.fit(nw_input, nw_output, verbose=2, batch_size=batch_size)

                my_acc = np.mean(np.round(model.predict(nw_input)) == nw_output)
                print("MyBinaryAccuracy:                              {:.4f}".format(my_acc))

The interesting thing is now the generated output:
```
Epoch 1/1
2s - loss: 0.0772 - binary_accuracy: 1.0000
MyBinaryAccuracy:                              0.3125
```

As may be seen the "binary_accuracy" is completely different from my calculated binary accuracy. Another interesting fact is that "model.evaluate(nw_input, nw_output)" returns "[0.702..., 0.3125]". This looks right.

I use tensorflow as backend. Maybe someone has an idea what I am doing wrong?

Thank you very much

**Comment by kutoga** [comment]
Unfortunately the linked thread did not help me. I do not the "ImageDataGenerator", but only plain models (there was also another thread linked, but it did not help).

I have create a new more minimal GAN example that exactly has the same problem: I use the evalute method before and after the fit-method to get an idea what the binary_accuracy is about:

```
~~~~~~~~~~~~~~~~~~~~~~~~
Iteration 26
~~~~~~~~~~~~~~~~~~~~~~~~

Train discriminator
Binary accuracy: 1.0
Epoch 1/1
0s - loss: 4.7202e-07 - binary_accuracy: 1.0000
Binary accuracy: 1.0

Train generator
Binary accuracy: 0.0 ################## Output of evaluate method
Epoch 1/1
0s - loss: 1.1921e-07 - binary_accuracy: 1.0000 ##################  Output of fit method
Binary accuracy: 0.0 ##################  Output of evaluate method
```

As can be seen the binary_accuracy values from the evalute method are "0.0", but the fit-method says it is "1.0".

Maybe I do something completely wrong, but I thought that evaluate should return the same binary_crossentropy as the fit-method? I attachted the code below.

I would be very happy if someone could provide me a hint:)

Thank you very much

```
from keras.models import Model
from keras.layers import Dense, Input
from keras.layers import Reshape, Dropout
from keras.layers.core import Activation
from keras.layers.normalization import BatchNormalization
from keras.layers.convolutional import UpSampling2D
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.layers.core import Flatten
from keras.layers.noise import GaussianNoise
from keras.datasets import mnist
import numpy as np

data_format = "channels_first"
encoding_dim = 2
nonlinearity = 'relu'

def generator_model():
    input_layer = Input((encoding_dim,))

    nw = Dense(input_dim=100, units=1024)(input_layer)
    nw = Activation(nonlinearity)(nw)
    nw = Dense(128*7*7)(nw)
    nw = BatchNormalization()(nw)
    nw = Activation(nonlinearity)(nw)
    nw = Reshape((128, 7, 7), input_shape=(128*7*7,))(nw)
    nw = UpSampling2D(size=(2, 2), data_format=data_format)(nw)
    nw = Convolution2D(64, (5, 5), padding='same', data_format=data_format)(nw)
    nw = Activation(nonlinearity)(nw)
    nw = UpSampling2D(size=(2, 2), data_format=data_format)(nw)
    nw = Convolution2D(1, (5, 5), padding='same', data_format=data_format)(nw)
    nw = Activation('sigmoid')(nw)
    model = Model([input_layer], [nw], name="generator")
    return model


def discriminator_model(noisy=False):
    input_layer = Input((1, 28, 28))

    # Add noise if required:
    # Normalize the data before adding the noise, do this also if no noise is used. This allows it to reuse the
    # weights independent of if the network is trained with or without noise.
    nw = BatchNormalization()(input_layer)
    if noisy:
        nw = GaussianNoise(0.5)(nw)
        nw = Dropout(0.5)(nw)

    nw = Convolution2D(
                        64, (5, 5),
                        data_format=data_format,
                        padding='same')(nw)
    nw = Activation(nonlinearity)(nw)
    nw = MaxPooling2D(pool_size=(2, 2), data_format=data_format)(nw)
    nw = Convolution2D(128, (5, 5), data_format=data_format)(nw)
    nw = Activation(nonlinearity)(nw)
    nw = MaxPooling2D(pool_size=(2, 2), data_format=data_format)(nw)
    nw = Flatten()(nw)
    nw = Dense(1024)(nw)
    nw = Activation(nonlinearity)(nw)
    nw = Dense(1)(nw)
    nw = Activation('sigmoid')(nw)
    model = Model([input_layer], [nw], name="discriminator")
    return model

def generator_containing_discriminator(generator, discriminator):
    input_layer = Input((encoding_dim,))
    nw = generator(input_layer)
    nw = discriminator(nw)
    model = Model([input_layer], [nw], name="generator_containing_discriminator")
    return model

(X_train, y_train), (X_test, y_test) = mnist.load_data()
def org_img_selector(count):
    data = X_train[np.random.randint(X_train.shape[0], size=count)].reshape((count, 1, 28, 28))

    # Normalize the data into the range [0, 1]
    data = data.astype(np.float32) / 255

    return data

def noise(count):
    return np.random.uniform(0, 1, (count, encoding_dim))

# Create models
generator = generator_model()
discriminator = discriminator_model()
generator_discriminator = generator_containing_discriminator(generator, discriminator)

discriminator.compile(
    loss='binary_crossentropy',
    optimizer='adam',#SGD(lr=0.0005, momentum=0.9, nesterov=True),
    metrics=['binary_accuracy']
)
generator_discriminator.compile(
    loss='binary_crossentropy',
    optimizer='adam',#SGD(lr=0.0005, momentum=0.9, nesterov=True),
    metrics=['binary_accuracy']
)

# Start training
itr = 0
num_elements = 2048
while True:
    print("~~~~~~~~~~~~~~~~~~~~~~~~")
    print("Iteration {}".format(itr))
    print("~~~~~~~~~~~~~~~~~~~~~~~~")
    print()

    # Train the discriminator
    X_org = org_img_selector(num_elements // 2)
    y_org = np.ones((num_elements // 2,), dtype=np.float32)
    X_gen = generator.predict(noise(num_elements // 2))
    y_gen = np.zeros((num_elements // 2,), dtype=np.float32)
    X = np.concatenate([X_org, X_gen], axis=0)
    y = np.concatenate([y_org, y_gen], axis=0)

    print("Train discriminator")
    print("Binary accuracy: {}".format(discriminator.evaluate(X, y, verbose=0)[1]))
    discriminator.fit(X, y, batch_size=num_elements, verbose=2)
    print("Binary accuracy: {}".format(discriminator.evaluate(X, y, verbose=0)[1]))
    print()

    # Pretrain the discriminator for 10 episodes
    if itr > 10:

        # Train the generator
        X = noise(num_elements)
        y = np.ones((num_elements,))

        discriminator.trainable = False

        print("Train generator")
        print("Binary accuracy: {}".format(generator_discriminator.evaluate(X, y, verbose=0)[1]))
        generator_discriminator.fit(X, y, batch_size=num_elements, verbose=2)
        print("Binary accuracy: {}".format(generator_discriminator.evaluate(X, y, verbose=0)[1]))
        print()

        discriminator.trainable = True


    print()
    itr += 1


```
